### DISCUSSION TITLE
WebAssembly

### DISCUSSION THREAD

**Discussion by derekgdicamillo** [oqtane/oqtane.framework#2484] [Q&A]
I must be missing something. 

I have tried changing runtime to WebAssembly. I get the configuration page and fill it out and click install now and nothing happens. I did try it with server in runtime and it worked perfect. Is there something I may be missing?

**Comment by markdav-is** [comment]
I'm investigating a WASM implementation of an Oqtane project at the moment.  We need a mobile-first experience and mobile devices put the browser app to sleep causing blazor server to show the "connection lost" page.  I'm also looking at the MAUI support, as that might be a good path to mobile-first as well.  my understanding is that there's a lot of prep to pull off the WASM deployment style of Oqtane.  There are decisions to be made for what assemblies to include for example due to the dynamic nature.

**Comment by sbwalker** [comment]
@derekgdicamillo You did not indicate which version of Oqtane you are using... and this is important to know because the mechanics of how Server and WebAssembly is supported has changed over time (as Blazor itself has evolved over time). 

Originally it was using conditional compilation where you needed to choose a specific build profile in Visual Studio and build the framework for either Server or WebAssembly. Then it moved away from conditional compilation and was able to support both models in the same deployment... and you were able to vary your hosting model by setting a property in the appsettings.json - however this meant that the entire application could only run under a single hosting model. In Oqtane 3.0 improvements were added to support multiple Blazor hosting models within the same run-time deployment (https://www.oqtane.org/blog/!/32/announcing-oqtane-3-0-for-net-6). Basically, the hosting model support works in conjunction with the multi-tenant capabilities and allows you to specify the hosting model for each site in your deployment (ie. Site1 can run on Blazor Server, and Site2 on Blazor WebAssembly, etc...). 


![image](https://user-images.githubusercontent.com/4840590/200343220-1e30746d-607b-4dc6-9d79-4d7fc94051a8.png)

The legacy method of specifying the Runtime property in appsettings.json should still be supported but I will need to investigate further. In the meantime, install the framework on Blazor Server and then go into Site Settings in the Admin Dashboard and change the Runtime field to Blazor WebAssembly. The browser should refresh when you click Save... and the assemblies will be automatically downloaded to the browser from the server and you will be running on Blazor WebAssembly.

**Comment by sbwalker** [comment]
@markdav-is a feature was included in the 3.2.1 release (https://www.oqtane.org/blog/!/52/oqtane-3-2-1-released) to better handle the "connection lost" scenario on Blazor Server:

"Blazor Server uses SignalR to maintain a persistent connection between the client and server. However in order to preserve server resources, SignalR will disconnect the connection after a period of idle user activity. This results in a message being displayed to the user once they become active again of "Could not reconnect to the server. Reload the page to restore functionality.". This is not an optimal user experience, so in Oqtane 3.2.1 some logic was added to enable the client browser to automatically reconnect to the server when this scenario is encountered."

Is this not working in your scenario - or are you still using an older version?
